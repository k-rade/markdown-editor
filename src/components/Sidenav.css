.sidenav {
    width: 5rem;
    height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 1.5rem;
    padding: 0 1rem;
    border-right: 1px solid var(--color-gray-600);
}

.logo {
    height: 5rem;
    display: flex;
    align-items: center;
    font-weight: 600;
    font-size: 1.25rem;
    border-bottom: 1px solid var(--color-gray-600);
}

.tabs {
    gap: 1rem;
    display: flex;
    flex-direction: column;
}

.tab-active .icon>path {
    stroke: var(--color-white-one);
}

.tab-active {
    position: relative;
}

.tab-active::before {
    position: absolute;
    content: "";
    width: 2px;
    height: 100%;
    background-color: var(--color-primary-500);
    top: 0;
    right: -18px;
}

.files-button-container {
    display: none;
    position: relative;
}

.files-button-badge {
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: 5rem;
    font-size: 0.75rem;
    font-weight: 500;
    top: 4px;
    right: 0px;
    position: absolute;
    width: 1.25rem;
    height: 1.25rem;
    background-color: var(--color-primary-500);
    color: var(--color-black-one);
}

@media screen and (max-width: 1024px) {

    .files-button-container {
        display: block;
    }

    .logo {
        display: none;
    }

    .sidenav {
        bottom: 0;
        z-index: 1;
        border: 0;
        width: 100vw;
        height: 4rem;
        background-color: var(--color-gray-100);
        border-top: 1px solid var(--color-gray-500);
        position: fixed;
        justify-content: center;
        flex-direction: row;
    }

    .tab-active::before {
        width: 100%;
        height: 2px;
        background-color: var(--color-primary-500);
        top: -13px;
        left: 0;
    }
}