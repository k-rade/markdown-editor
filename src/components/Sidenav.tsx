import { useState } from "react";
import { FiX, FiMenu } from "react-icons/fi";
import { VscColorMode } from "react-icons/vsc";
import { Tab } from "../types/markdown";
import { MdModeEdit, MdFolder } from "react-icons/md";
import { AiFillEye } from "react-icons/ai";
import { IoIosDownload } from "react-icons/io";
import { TiDelete } from "react-icons/ti";

interface Props {
  activeTab: Tab;
  changeTab: (tab: Tab) => void;
  clearCurrentFile: () => void;
  downloadFile: () => void;
  toggleShowFiles: () => void;
}

const Sidenav = ({
  activeTab,
  changeTab,
  clearCurrentFile,
  downloadFile,
  toggleShowFiles,
}: Props): JSX.Element => {
  const [showMenu, setShowMenu] = useState(false);

  const handleMenuOpen = (): void => {
    setShowMenu(true);
  };

  const handleMenuClose = (): void => {
    setShowMenu(false);
  };

  const handleTabChange = (tab: Tab): void => {
    changeTab(tab);
    handleMenuClose();
  };

  const handleFileClear = (): void => {
    clearCurrentFile();
    handleMenuClose();
  };

  const handleFileDownload = (): void => {
    downloadFile();
    handleMenuClose();
  };

  const handleFilesShowToggle = (): void => {
    toggleShowFiles();
    handleMenuClose();
  };

  const handleDarkThemeSwitch = (): void => {
    if (document.body.classList.contains("dark")) {
      document.body.classList.remove("dark");
    } else {
      document.body.classList.add("dark");
    }
  };

  return (
    <>
      {showMenu && (
        <div
          data-testid="sidenav-overlay"
          onClick={handleMenuClose}
          className="z-10 sm:hidden fixed inset-0 w-full h-full bg-neutral-900/90 z-100"
        ></div>
      )}
      <div className="overflow-y-auto w-full sm:max-w-[5rem] bg-neutral-50 dark:bg-neutral-800 flex flex-col items-center gap-4 sm:border-r dark:border-r-neutral-700">
        <div className="bg-neutral-50 dark:bg-neutral-800 p-4 justify-between flex items-center sm:justify-center sm:px-0 w-full z-10">
          <div className="w-8 h-8 border-neutral-800 dark:border-white border-8 rounded-full"></div>
          {showMenu ? (
            <FiX
              data-testid="x-icon"
              onClick={handleMenuClose}
              size="1.5rem"
              className="stroke-neutral-800 dark:stroke-white sm:hidden"
            />
          ) : (
            <FiMenu
              data-testid="menu-icon"
              onClick={handleMenuOpen}
              size="1.5rem"
              className="stroke-neutral-800 dark:stroke-white sm:hidden"
            />
          )}
        </div>
        <div
          data-testid="sidenav-menu"
          className={`${
            showMenu ? "top-0" : "-top-full"
          } sm:h-full sm:bg-none z-10 transition-[top] duration-300 absolute sm:relative sm:mt-0 py-2 sm:py-0 sm:top-0 mt-16 bg-neutral-50 dark:bg-neutral-800 flex flex-col items-center gap-1 sm:gap-4 w-full px-2 sm:px-4`}
        >
          <div
            onClick={(): void => handleTabChange("editor")}
            className={`${
              activeTab === "editor"
                ? "bg-neutral-200 dark:bg-neutral-700/50"
                : ""
            } hover:cursor-pointer hover:bg-neutral-200 dark:hover:bg-neutral-700/50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full`}
          >
            <MdModeEdit
              data-testid="edit-icon"
              size="1.5rem"
              className="fill-neutral-500"
            />
            <span className="text-neutral-800 dark:text-neutral-50 sm:hidden">
              Editor
            </span>
          </div>
          <div
            onClick={(): void => handleTabChange("preview")}
            className={`${
              activeTab === "preview"
                ? "bg-neutral-200 dark:bg-neutral-700/50"
                : ""
            } hover:cursor-pointer hover:bg-neutral-200 dark:hover:bg-neutral-700/50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full`}
          >
            <AiFillEye
              data-testid="eye-icon"
              size="1.5rem"
              className="fill-neutral-500"
            />
            <span className="text-neutral-800 dark:text-neutral-50 sm:hidden">
              Preview
            </span>
          </div>
          <div
            onClick={handleFileDownload}
            className="hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <IoIosDownload
              data-testid="download-icon"
              size="1.5rem"
              className="fill-neutral-500"
            />
            <span className="text-neutral-800 dark:text-neutral-50 sm:hidden">
              Download
            </span>
          </div>
          <div
            onClick={handleFileClear}
            className="hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <TiDelete
              data-testid="x-circle-icon"
              size="1.5rem"
              className="fill-neutral-500"
            />
            <span className="text-neutral-800 dark:text-neutral-50 sm:hidden">
              Clear
            </span>
          </div>
          <div
            onClick={handleFilesShowToggle}
            className="lg:hidden hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <MdFolder
              data-testid="folder-icon"
              size="1.5rem"
              className="fill-neutral-500"
            />
            <span className="text-neutral-800 dark:text-neutral-50 sm:hidden">
              Files
            </span>
          </div>
          <div className="hidden sm:block bg-neutral-200 dark:bg-neutral-700 h-px w-8"></div>
          <div
            onClick={handleDarkThemeSwitch}
            className="hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <VscColorMode
              data-testid="theme-switch-icon"
              size="1.125rem"
              className="fill-neutral-500"
            />
            <span className="text-neutral-800 dark:text-neutral-50 sm:hidden">
              Theme
            </span>
          </div>
        </div>
      </div>
    </>
  );
};

export default Sidenav;
