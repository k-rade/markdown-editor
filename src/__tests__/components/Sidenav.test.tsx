import { fireEvent, render, RenderResult } from "@testing-library/react";
import Sidenav from "../../components/Sidenav";

describe("Sidenav", (): void => {
  const mockChangeTab = jest.fn();
  const mockClearCurrentFile = jest.fn();
  const mockDownloadFile = jest.fn();
  const mockToggleShowFiles = jest.fn();

  const renderTestComponent = (): RenderResult =>
    render(
      <Sidenav
        activeTab="editor"
        changeTab={mockChangeTab}
        clearCurrentFile={mockClearCurrentFile}
        downloadFile={mockDownloadFile}
        toggleShowFiles={mockToggleShowFiles}
      />
    );

  beforeEach((): void => {
    jest.clearAllMocks();
  });

  it("should render", (): void => {
    const { baseElement } = renderTestComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it("should open menu", (): void => {
    const { getByTestId } = renderTestComponent();

    fireEvent.click(getByTestId("menu-icon"));
    expect(getByTestId("sidenav-overlay")).toBeInTheDocument();
    expect(getByTestId("x-icon")).toBeInTheDocument();
    expect(getByTestId("sidenav-menu").className).toContain("top-0");
  });

  it("should close menu when clicked on x icon", (): void => {
    const { getByTestId, queryByTestId } = renderTestComponent();

    fireEvent.click(getByTestId("menu-icon"));
    fireEvent.click(getByTestId("x-icon"));

    expect(queryByTestId("sidenav-overlay")).not.toBeInTheDocument();
    expect(queryByTestId("x-icon")).not.toBeInTheDocument();
    expect(getByTestId("sidenav-menu").className).toContain("-top-full");
  });

  it("should close menu when clicked on sidenav overlay", (): void => {
    const { getByTestId, queryByTestId } = renderTestComponent();

    fireEvent.click(getByTestId("menu-icon"));
    fireEvent.click(getByTestId("sidenav-overlay"));

    expect(queryByTestId("sidenav-overlay")).not.toBeInTheDocument();
    expect(queryByTestId("x-icon")).not.toBeInTheDocument();
    expect(getByTestId("sidenav-menu").className).toContain("-top-full");
  });

  it("should toggle files show", (): void => {
    const { getByTestId } = renderTestComponent();

    fireEvent.click(getByTestId("folder-icon"));

    expect(mockToggleShowFiles).toHaveBeenCalled();
  });

  it("should download file", (): void => {
    const { getByTestId } = renderTestComponent();

    fireEvent.click(getByTestId("download-icon"));

    expect(mockDownloadFile).toHaveBeenCalledTimes(1);
  });

  it("should clear current file", (): void => {
    const { getByTestId } = renderTestComponent();

    fireEvent.click(getByTestId("x-circle-icon"));

    expect(mockClearCurrentFile).toHaveBeenCalledTimes(1);
  });

  it.each([
    ["editor", "edit-icon"],
    ["preview", "eye-icon"],
  ])(
    "should change active tab to %s when clicked on %s",
    (tabName: string, icon: string): void => {
      const { getByTestId } = renderTestComponent();

      fireEvent.click(getByTestId(icon));

      expect(mockChangeTab).toHaveBeenNthCalledWith(1, tabName);
    }
  );
});
