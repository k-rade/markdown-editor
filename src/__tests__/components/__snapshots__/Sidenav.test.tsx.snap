// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidenav should render 1`] = `
<body>
  <div>
    <div
      class="overflow-y-auto w-full sm:max-w-[5rem] bg-neutral-50 dark:bg-neutral-800 flex flex-col items-center gap-4 sm:border-r dark:border-r-neutral-700"
    >
      <div
        class="bg-neutral-50 dark:bg-neutral-800 p-4 justify-between flex items-center sm:justify-center sm:px-0 w-full z-10"
      >
        <div
          class="w-8 h-8 border-neutral-800 dark:border-white border-8 rounded-full"
        />
        <svg
          class="stroke-neutral-800 dark:stroke-white sm:hidden"
          data-testid="menu-icon"
          fill="none"
          height="1.5rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="3"
            x2="21"
            y1="12"
            y2="12"
          />
          <line
            x1="3"
            x2="21"
            y1="6"
            y2="6"
          />
          <line
            x1="3"
            x2="21"
            y1="18"
            y2="18"
          />
        </svg>
      </div>
      <div
        class="-top-full sm:h-full sm:bg-none z-10 transition-[top] duration-300 absolute sm:relative sm:mt-0 py-2 sm:py-0 sm:top-0 mt-16 bg-neutral-50 dark:bg-neutral-800 flex flex-col items-center gap-1 sm:gap-4 w-full px-2 sm:px-4"
        data-testid="sidenav-menu"
      >
        <div
          class="bg-neutral-200 dark:bg-neutral-700/50 hover:cursor-pointer hover:bg-neutral-200 dark:hover:bg-neutral-700/50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
        >
          <svg
            class="fill-neutral-500"
            data-testid="edit-icon"
            fill="currentColor"
            height="1.5rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <span
            class="text-neutral-800 dark:text-neutral-50 sm:hidden"
          >
            Editor
          </span>
        </div>
        <div
          class=" hover:cursor-pointer hover:bg-neutral-200 dark:hover:bg-neutral-700/50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
        >
          <svg
            class="fill-neutral-500"
            data-testid="eye-icon"
            fill="currentColor"
            height="1.5rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M396 512a112 112 0 1 0 224 0 112 112 0 1 0-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 0 0 0 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
            />
          </svg>
          <span
            class="text-neutral-800 dark:text-neutral-50 sm:hidden"
          >
            Preview
          </span>
        </div>
        <div
          class="hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
        >
          <svg
            class="fill-neutral-500"
            data-testid="download-icon"
            fill="currentColor"
            height="1.5rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M376 144H269v179.9l47.9-47.2c5.1-5 13.3-5 18.4.1 5 5.1 5 13.3-.1 18.4l-70 69c-2.5 2.4-5.8 3.7-9.1 3.7-1.7 0-3.4-.3-5-1-1.5-.6-2.9-1.6-4.1-2.7l-70-69c-5.1-5-5.2-13.3-.1-18.4 5-5.1 13.3-5.2 18.4-.1l47.9 47.2V144H136c-22 0-40 18-40 40v240c0 22 18 40 40 40h240c22 0 40-18 40-40V184c0-22-18-40-40-40zM269 61c0-7.2-5.8-13-13-13s-13 5.8-13 13v83h26V61z"
            />
          </svg>
          <span
            class="text-neutral-800 dark:text-neutral-50 sm:hidden"
          >
            Download
          </span>
        </div>
        <div
          class="hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
        >
          <svg
            baseProfile="tiny"
            class="fill-neutral-500"
            data-testid="x-circle-icon"
            fill="currentColor"
            height="1.5rem"
            stroke="currentColor"
            stroke-width="0"
            version="1.2"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4c-4.419 0-8 3.582-8 8s3.581 8 8 8 8-3.582 8-8-3.581-8-8-8zm3.707 10.293c.391.391.391 1.023 0 1.414-.195.195-.451.293-.707.293s-.512-.098-.707-.293l-2.293-2.293-2.293 2.293c-.195.195-.451.293-.707.293s-.512-.098-.707-.293c-.391-.391-.391-1.023 0-1.414l2.293-2.293-2.293-2.293c-.391-.391-.391-1.023 0-1.414s1.023-.391 1.414 0l2.293 2.293 2.293-2.293c.391-.391 1.023-.391 1.414 0s.391 1.023 0 1.414l-2.293 2.293 2.293 2.293z"
            />
          </svg>
          <span
            class="text-neutral-800 dark:text-neutral-50 sm:hidden"
          >
            Clear
          </span>
        </div>
        <div
          class="lg:hidden hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
        >
          <svg
            class="fill-neutral-500"
            data-testid="folder-icon"
            fill="currentColor"
            height="1.5rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
            />
          </svg>
          <span
            class="text-neutral-800 dark:text-neutral-50 sm:hidden"
          >
            Files
          </span>
        </div>
        <div
          class="hidden sm:block bg-neutral-200 dark:bg-neutral-700 h-px w-8"
        />
        <div
          class="hover:bg-neutral-200 dark:hover:bg-neutral-700/50 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
        >
          <svg
            class="fill-neutral-500"
            data-testid="theme-switch-icon"
            fill="currentColor"
            height="1.125rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1.125rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1a7 7 0 1 0 0 14A7 7 0 0 0 8 1zm0 13V2a6 6 0 1 1 0 12z"
            />
          </svg>
          <span
            class="text-neutral-800 dark:text-neutral-50 sm:hidden"
          >
            Theme
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
